### ISSUE TITLE
Possible feature request? Collections that save the module contents instead of simply a list of modules to be reloaded

### ISSUE THREAD

**Issue by damianam** [TACC/Lmod#219]
Context: we might need to support an use case where the modules have to be loaded independently by each process, as opposed as the master process in the job loading the module and forwarding the environment to all the job processes (we use slurm).

In this context, we might go from `O(m)` to `O(m*p)` (with `m` being the number of modules to be loaded and `p` being the total number of processes, being `m<<p`) accesses to the filesystem that holds the modules.

If there were some kind of collection that saved not the module table, but the commands issued by the listed modules, we could transform `O(m*p)` into `O(p)`, which could be an improvement of up 30 (in terms of files accessed) in some cases.

Is that something doable and interesting? At first sight, it might be, particularly for people installing software in shared filesystems.

**Comment by rtmclay** [comment]
There is nothing like this currently.  There is also a problem a collection of commands being out-of-date with the underlying modulefiles.

You can do this now however with a little bit of scripting:

Convert the module commands from module to "$LMOD_CMD bash ..."

```
   $    $LMOD_CMD bash load a b c > myEnv.sh
   $    $LMOD_CMD bash load d e >> myEnv.sh

```

Then push myEnv.sh to all processes.

**Comment by damianam** [comment]
I think that the risk of having an outdated collection of commands is worthy it. We generally do not change module logic in production, so that'd be a very special case. The probability of an user using that feature times the probability of having a bug in a module that requires an update should be very small.

That's a neat scripting trick, but I am not sure it is production ready. Users will find a way to mess it up, it would be easy to end up sourcing `myEnv2_test.sh` instead of `myEnv.sh`, particularly if they start using that to put more stuff in there to set up their environment manually.

**Comment by rtmclay** [comment]
Supporting this would take a lot of work which I don't have the time.   Also I'm concerned about the following: While you aren't worried about this new collection out-of-date with the underlying modules, I have to be.  

Also, in my experience, this isn't going to work.  Essentially, you want to replace m modules with 1 module.  If the number of tasks (p) is high, say 4096 nodes, this might fail.  We have seen lustre fail at that size.   Besides the most of the cost of loading a module, is the requirement to walk all of the module tree.  This happens once when one does "module load a b c d".    You should measure the time on your system of loading one module versus several.   Module load time decrease a great deal when using LMOD_CACHED_LOADS=1.  (with the problem of keeping the cache file up-to-date).

Another test you might do is create a single module out of a group of modules.  You can use the sh_to_modulefile.sh script to build a single modulefile and test that to see if it helps.

You haven't made clear the use case where you need this so I can't comment on that.   But I can say that we have made this difficult for users to do this on our systems.   A user might try to have a job script:

      ibrun my_shell_script.sh

where ibrun is our command to run in parallel.  The user is running a shell script in parallel.   We to allow that.  What we done is turn off the module command.   That is the command runs but does nothing.  So if there are any module commands in "my_shell_script.sh" they do nothing.

**Comment by damianam** [comment]
Your observation is right, and that's part of an internal discussion we are having. So we aren't sure if this will become a pressing issue for us.

But regardless, in certain environments you can be easily loading 20+ modules each time. Think of EasyBuild setups, with the compiler, MPI, libraries and build tools that are loaded every time. In that case, you can replace 20+ loads with 1. Sure, with the cache you avoid walking the tree, which helps a lot already, but currently you can't avoid reading 20+ modules. If the file system fails with 4096 nodes loading a couple of modules, it would take a fraction of it to fail when loading 20+ modules per process. This kind of thing wouldn't solve the issue, but it would move up the number of tasks where it becomes a problem.

I can't comment much about the use case at this point, but the motivation is basically that we will need to spawn jobs that will contain 2 or more groups of processes with different environment each group. With SLURM, the environment is loaded in the master process and then forwarded to the other processes, but to support this kind of use case we need to reload the environment on each process. We can think too of modifying SLURM to allow multiple 1 master process per group of processes, instead of 1 per job, if that is more robust solution.

**Comment by rtmclay** [comment]
O.K. to close this issue?

**Comment by damianam** [comment]
Sorry for the late answer, I was offline. I personally wouldn't close it, unless you consider it a "won't fix".

**Comment by rtmclay** [comment]
This a won't fix issue for the reasons I stated above.  This can't be made safe.

**Comment by fgeorgatos** [comment]
Hello folks,

Sry for the late hop in, but this rings a bell.

Would a one-module-load solution be of interest?! If yes, there is a middle
ground which is akin to what I have been calling stowed-toolchains or
-bundles. Basically, there is a GNU tool called Stow (its in EB btw) which
creates a symlink farm out of a bunch of directories. It would cut down on
env changes drastically (a worry with fat bundles). Damian, would that fit
your use case?!

F
-- 
echo "sysadmin know better bash than english"|sed s/min/mins/ \
  | sed 's/better bash/bash better/' # signal detected in a CERN forum
